iroh-util/src/lock.rs

```diff
@@ -1,79 +1,304 @@
-/// or when the program stops.
+pub fn acquire_or_exit(lock: &mut ProgramLock) -> &mut ProgramLock {
+    match lock.is_locked() {
+        Ok(false) => {
+            if let Err(e) = lock.acquire() {
+                eprintln!("error locking {}: {}", lock.program_name(), e);
+                std::process::exit(crate::exitcodes::ERROR);
+            }
+            lock
+        }
+        Ok(true) => {
+            eprintln!("{} is already running, stopping.", lock.program_name());
+            std::process::exit(crate::exitcodes::LOCKED);
+        }
+        Err(err) => {
+            eprintln!("error checking lock {}: {}", lock.program_name(), err);
+            std::process::exit(crate::exitcodes::ERROR);
+        }
+    }
+}
+
+/// Manages a lock file used to track if an iroh program is already running.
+/// Aquired locks write a file to iroh's application data path containing the
+/// process identifier (PID) of the process with the lock.
+/// The lock exclusion test requires both a lockfile AND a running process
+/// listed at the PID in the file
+/// An acquired lock is released either when the object is dropped
+/// or when the program stops, which removes the file
+/// Invalid or corrupt locks are overwritten on acquisition
 pub struct ProgramLock {
     path: PathBuf,
-    lock: Option<FileGuard<Rc<File>>>,
+    lock: Option<sysinfo::Pid>,
 }
 
 impl ProgramLock {
     /// Create a new lock for the given program. This does not yet acquire the lock.
-    pub fn new(prog_name: &str) -> Result<Self> {
-        let path = crate::iroh_data_path(&format!("{}.lock", prog_name))?;
+    pub fn new(prog_name: &str) -> Result<Self, LockError> {
+        let path = crate::iroh_data_path(&format!("{}.lock", prog_name))
+            .map_err(|e| LockError::InvalidPath { source: e })?;
         Ok(Self { path, lock: None })
     }
 
+    /// Shorthand intended for main functions that need a lock to guard the process
+    pub fn acquire_or_exit(&mut self) -> &mut Self {
+        match self.is_locked() {
+            Ok(false) => {
+                if let Err(e) = self.acquire() {
+                    eprintln!("error locking {}: {}", self.program_name(), e);
+                    std::process::exit(crate::exitcodes::ERROR);
+                }
+                self
+            }
+            Ok(true) => {
+                eprintln!("{} is already running, stopping.", self.program_name());
+                std::process::exit(crate::exitcodes::LOCKED);
+            }
+            Err(err) => {
+                eprintln!("error checking lock {}: {}", self.program_name(), err);
+                std::process::exit(crate::exitcodes::ERROR);
+            }
+        }
+    }
+
+    pub fn path(&self) -> &PathBuf {
+        &self.path
+    }
+
+    pub fn program_name(&self) -> &str {
+        self.path
+            .file_name()
+            .unwrap_or_else(|| std::ffi::OsStr::new(""))
+            .to_str()
+            .unwrap()
+            .split('.')
```

Comment:
Why do you need to split on `.` here?

Reply:
`.lock` extension in the file name. There's a forthcoming method in rust nightly that'll make this much cleaner